d1985r3.md

```diff
@@ -376,148 +303,114 @@ limitations we saw before `if constexpr` was introduced: Function templates
 often had to rely on helper structs to do simple things.
 
 
-### Allowing unversal template parameters in code
+### Dependent Names
 
 To make this feature easier to use we also propose that universal template
 parameters are usable in code. To make parsing possible we then have to
 disambiguate universal template parameters according to dependent name rules.
 
-Unfortunately always disambiguating as value prevents `using type =
-X<U>;` above from working as expected. U would be automatically disambiguated as
-a value at parse time and if it turns out not to be a value at instantiation
-time this is an error.
+The dependent name rules shall be modified such that dependent names passed as
+template arguments are promoted as universal template parameters. This allows
+`using type = X<U>;` above to work as expected and slighlty extends "Down with
+`typename`!".
 
-We actually have a version of this problem already with overloaded template
-functions and dependent names:
+This rule shall apply to dependent names other than universal template
+parameters, as well:
 
 ```cpp
 template<typename T> int f() { return sizeof(T); }
 template<auto V> int f() { return V; }
 
 template<typename T> int caller()
 {
-    return f<??? T::name>();
+    return f<T::name>();
 }
 ```
 
-Here the second f is called if T::name is a value, but if it is a type an
-instantiation error is caused, rather than the first f being called.
+Because `T::name` is a dependent type used as a template argument, it is
+promoted to a universal template parameter and the correct `f()` is called.
 
-With universal template parameters this problem will be more articulated,
-especially when universal template parameters are forwarded directly as in:
+This rule also applies when universal template parameters are not explicitly
+used, and thus slightly extends "Down with `typename`!".
 
 ```cpp
-template<template auto U> int caller2()
-{
-    return f<U>();
-}
+std::vector<T::name> v;
 ```
 
-This compiles with C++20 disambiguation rules but always tries to call the
-second overload of f as U is always treated as a value, with the surprising result
-that `caller2` can't be instantiated with a type even though f has an overload
-for types.
+In C++20, `T::name` would be parsed as a value. However, according to this new
+rule, `T::name` would become a universal template parameter, which would then
```

Comment:
I don't think it becomes a universal template parameter, rather dependent names and universal template parameters work the same.

Reply:
Look at it in non-diff format. This is a rewrite of both our sections.

We need to show how it affects dependent names. I think it makes a lot of sense to treat these dependent names as promoted to UTPs, the "root" of all types/NTTPs/etc. that can then become narrowed in context.